components/browser/session/src/main/java/mozilla/components/browser/session/SessionManager.kt

```diff
@@ -318,9 +328,55 @@ class SessionManager(
      * Informs this [SessionManager] that the OS is in low memory condition so it
      * can reduce its allocated objects.
      */
+    @Deprecated("Use onTrimMemory instead.", replaceWith = ReplaceWith("onTrimMemory"))
     fun onLowMemory() {
-        delegate.onLowMemory()
-        store?.syncDispatch(SystemAction.LowMemoryAction)
+        onTrimMemory(ComponentCallbacks2.TRIM_MEMORY_MODERATE)
+    }
+
+    @Synchronized
+    @Suppress("NestedBlockDepth")
+    override fun onTrimMemory(level: Int) {
+        val clearThumbnails = level >= ComponentCallbacks2.TRIM_MEMORY_RUNNING_LOW
+        val closeEngineSessions = level >= ComponentCallbacks2.TRIM_MEMORY_RUNNING_CRITICAL
+
+        logger.debug("onTrimMemory($level): clearThumbnails=$clearThumbnails, closeEngineSessions=$closeEngineSessions")
+
+        if (!clearThumbnails && !closeEngineSessions) {
+            // Nothing to do for now.
+            return
+        }
+
+        val selectedSession = selectedSession
+        val states = mutableMapOf<String, EngineSessionState>()
+        val sessions = sessions
+
+        var clearedThumbnails = 0
+        var unlinkedEngineSessions = 0
+
+        sessions.forEach {
+            if (it != selectedSession) {
+                if (clearThumbnails) {
+                    it.thumbnail = null
+                    clearedThumbnails++
+                }
+
+                if (closeEngineSessions) {
+                    val state = it.engineSessionHolder.engineSession?.saveState()
+                    linker.unlink(it)
+
+                    if (state != null) {
+                        it.engineSessionHolder.engineSessionState = state
+                        states[it.id] = state
+                    }
+
+                    unlinkedEngineSessions++
+                }
+            }
+        }
+
+        logger.debug("Cleared $clearThumbnails thumbnail(s) and unlinked $unlinkedEngineSessions engine sessions()")
```

Comment:
Also nit: don't need to end sentence in `()` :)

Reply:
@jonalmeida doesn't like me to end my sentences with `.` - now you are telling me `()` is also not an option?? :)

Reply:
(That should have been `session(s)` - plurals, yay!)

Reply:
Only commit messages! 😅

Reply:
😆